File: greis/geodesy.py
"""Geodetic coordinates to earth-centred, earth-fixed ones.

The receiver reports where it is as latitude, longitude and height on the
WGS-84 ellipsoid. ECEF is the same point in a right-handed Cartesian frame
whose origin is the centre of mass of the earth, whose Z axis runs through
the reference pole, and whose X axis runs through the intersection of the
prime meridian and the equator. Nothing is estimated in getting from one
to the other: it is a closed-form transform, exact to the precision of the
arithmetic.

This is why ECEF is offered as a derived column rather than as another
message to ask the receiver for. GREIS can report Cartesian position
directly, but a receiver computing it from its own solution and this code
computing it from the same solution's latitude, longitude and height
produce the same numbers - so the extra message would cost bandwidth on
the serial link and buy nothing, and every session already recorded could
not be converted after the fact, which this can.

**The height has to be ellipsoidal.** GREIS [PG] reports height above the
WGS-84 ellipsoid, which is what this needs. A receiver configured to
output orthometric height instead - height above the geoid - would produce
an ECEF position wrong by the geoid separation, which around the
Mediterranean is on the order of twenty metres. That is a receiver
configuration this application does not read and deliberately does not
change; it is called out in the README rather than guessed at here.
"""

from __future__ import annotations

import math

WGS84_SEMI_MAJOR_AXIS_M = 6378137.0
"""``a``. The equatorial radius, exact by definition of WGS-84."""

WGS84_INVERSE_FLATTENING = 298.257223563
"""``1/f``. Defining constant of WGS-84, not a measurement."""

WGS84_FLATTENING = 1.0 / WGS84_INVERSE_FLATTENING

WGS84_ECCENTRICITY_SQUARED = WGS84_FLATTENING * (2.0 - WGS84_FLATTENING)
"""``e² = 2f - f²``. Derived rather than written out, so the ellipsoid is
defined once by its two defining constants and everything else follows."""

WGS84_SEMI_MINOR_AXIS_M = WGS84_SEMI_MAJOR_AXIS_M * (1.0 - WGS84_FLATTENING)
"""``b``. 6356752.314245... - not a round number, and not typed in as one."""


def prime_vertical_radius(latitude_rad: float) -> float:
    """``N``: the radius of curvature in the prime vertical.

    The distance from the point on the ellipsoid to the Z axis along the
    normal. It is what makes the transform an ellipsoid transform rather
    than a sphere one, and it is the only part of the formula that depends
    on the flattening.
    """
    sin_latitude = math.sin(latitude_rad)
    return WGS84_SEMI_MAJOR_AXIS_M / math.sqrt(
        1.0 - WGS84_ECCENTRICITY_SQUARED * sin_latitude * sin_latitude
    )


def geodetic_to_ecef(
    latitude_deg: float, longitude_deg: float, height_m: float
) -> tuple[float, float, float]:
    """``(X, Y, Z)`` in metres, from WGS-84 latitude, longitude and
    ellipsoidal height.

    Degrees in, because that is what the rest of this application carries -
    [PG] sends radians and the parser converts them once, at the edge, so
    that no two places in the codebase disagree about which unit a latitude
    is in.
    """
    latitude_rad = math.radians(latitude_deg)
    longitude_rad = math.radians(longitude_deg)

    sin_latitude = math.sin(latitude_rad)
    cos_latitude = math.cos(latitude_rad)
    radius = prime_vertical_radius(latitude_rad)

    x = (radius + height_m) * cos_latitude * math.cos(longitude_rad)
    y = (radius + height_m) * cos_latitude * math.sin(longitude_rad)
    # The (1 - e²) is the whole difference between a sphere and an
    # ellipsoid here: the Z axis is shortened by the flattening, which at
    # the pole is the 21 km between the semi-major and semi-minor axes.
    z = (radius * (1.0 - WGS84_ECCENTRICITY_SQUARED) + height_m) * sin_latitude
    return x, y, z

File: ios/GreisKit/Sources/GreisKit/Geodesy.swift
import Foundation

/// Geodetic coordinates to earth-centred, earth-fixed ones.
/// Ported from `greis/geodesy.py`.
///
/// The receiver reports where it is as latitude, longitude and height on
/// the WGS-84 ellipsoid. ECEF is the same point in a right-handed
/// Cartesian frame centred on the earth's centre of mass, with Z through
/// the reference pole and X through the intersection of the prime meridian
/// and the equator. Nothing is estimated in getting between them: it is a
/// closed-form transform, exact to the precision of the arithmetic.
///
/// That is why this is a derived column rather than another message to ask
/// the receiver for. GREIS can report Cartesian position directly, but the
/// receiver would be computing it from the same solution — so the extra
/// message would cost bandwidth and buy nothing, and no already-recorded
/// session could be converted afterwards, which this can.
///
/// **The height must be ellipsoidal.** GREIS [PG] reports height above the
/// WGS-84 ellipsoid, which is what this needs. A receiver configured for
/// orthometric height instead would produce an ECEF position wrong by the
/// geoid separation — around twenty metres in the eastern Mediterranean.
public enum Geodesy {

    /// `a`, the equatorial radius. Exact by definition of WGS-84.
    public static let semiMajorAxisM = 6_378_137.0

    /// `1/f`. A defining constant of WGS-84, not a measurement.
    public static let inverseFlattening = 298.257223563

    public static let flattening = 1.0 / inverseFlattening

    /// `e² = 2f - f²`. Derived, so the ellipsoid is defined once by its two
    /// defining constants and everything else follows from them.
    public static let eccentricitySquared = flattening * (2.0 - flattening)

    /// `b`. 6356752.314245…, computed rather than typed in.
    public static let semiMinorAxisM = semiMajorAxisM * (1.0 - flattening)

    /// `N`: the radius of curvature in the prime vertical — the distance
    /// from the point on the ellipsoid to the Z axis along the normal. It
    /// is the only part of the transform that depends on the flattening,
    /// and the only thing making it an ellipsoid transform rather than a
    /// spherical one.
    public static func primeVerticalRadius(latitudeRad: Double) -> Double {
        let sinLatitude = sin(latitudeRad)
        return semiMajorAxisM / (1.0 - eccentricitySquared * sinLatitude * sinLatitude).squareRoot()
    }

    /// `(x, y, z)` in metres, from WGS-84 latitude, longitude and
    /// ellipsoidal height in degrees and metres.
    ///
    /// Degrees in, because that is what the rest of the package carries:
    /// [PG] sends radians and the parser converts them once, at the edge,
    /// so no two places disagree about the unit of a latitude.
    public static func geodeticToECEF(
        latitudeDeg: Double, longitudeDeg: Double, heightM: Double
    ) -> (x: Double, y: Double, z: Double) {
        let latitudeRad = latitudeDeg * .pi / 180.0
        let longitudeRad = longitudeDeg * .pi / 180.0

        let sinLatitude = sin(latitudeRad)
        let cosLatitude = cos(latitudeRad)
        let radius = primeVerticalRadius(latitudeRad: latitudeRad)

        let x = (radius + heightM) * cosLatitude * cos(longitudeRad)
        let y = (radius + heightM) * cosLatitude * sin(longitudeRad)
        // The (1 - e²) is the whole difference between a sphere and an
        // ellipsoid here: it shortens the Z axis by the flattening, which
        // at the pole is the 21 km between the two axes.
        let z = (radius * (1.0 - eccentricitySquared) + heightM) * sinLatitude
        return (x, y, z)
    }
}
